packages/agreement-process/src/model/domain/agreement-validators.ts

```diff
@@ -149,77 +151,70 @@ export function assertRequesterIsProducer(
   agreement: Agreement,
   authData: AuthData,
   readModelService: ReadModelService
 ): Promise<void> => {
+  // This operation has a dedicated assertion because it's the only operation that
+  // can be performed also by the producer even when there is an active producer delegation
   try {
     assertRequesterIsConsumer(agreement, authData);
   } catch (error) {
     try {
       assertRequesterIsProducer(agreement, authData);
     } catch (error) {
-      const producerDelegation =
-        await readModelService.getDelegationByDelegateId(
-          authData.organizationId,
-          delegationKind.delegatedProducer
+      const activeProducerDelegation =
+        await readModelService.getActiveProducerDelegationByEserviceId(
+          agreement.eserviceId
         );
-      assertRequesterIsDelegate(producerDelegation?.delegateId, authData);
+      assertIsDelegateProducer(activeProducerDelegation, agreement, authData);
     }
   }
 };
 
-export const assertRequesterIsProducerOrDelegateProducer = (
+export const assertCanActAsProducer = async (
   agreement: Agreement,
-  delegateProducerId: TenantId | undefined,
-  authData: AuthData
-): void => {
-  if (delegateProducerId) {
-    assertRequesterIsDelegate(delegateProducerId, authData);
-  } else {
+  authData: AuthData,
+  activeProducerDelegation: Delegation | undefined
+): Promise<void> => {
+  if (!activeProducerDelegation) {
+    // No active producer delegation, requester is auhorized only if they are the producer
     assertRequesterIsProducer(agreement, authData);
+  } else {
+    // Active producer delegation, requester is authorized only if they are the delegate
+    assertIsDelegateProducer(activeProducerDelegation, agreement, authData);
   }
 };
 
-export const assertRequesterIsDelegate = (
-  delegateId: TenantId | undefined,
-  authData: AuthData
-): void => {
-  if (authData.organizationId !== delegateId) {
-    throw operationNotAllowed(authData.organizationId);
-  }
-};
-
-export const assertRequesterCanActivate = (
+const assertIsDelegateProducer = (
+  activeProducerDelegation: Delegation | undefined,
   agreement: Agreement,
-  delegateProducerId: TenantId | undefined,
   authData: AuthData
 ): void => {
-  try {
-    assertRequesterIsConsumer(agreement, authData);
-  } catch (e) {
-    if (delegateProducerId) {
-      assertRequesterIsDelegate(delegateProducerId, authData);
-    } else {
-      assertRequesterIsProducer(agreement, authData);
-    }
+  if (
+    activeProducerDelegation?.delegateId !== authData.organizationId ||
+    activeProducerDelegation?.delegatorId !== agreement.producerId ||
+    activeProducerDelegation?.kind !== delegationKind.delegatedProducer ||
+    activeProducerDelegation?.state !== delegationState.active ||
+    activeProducerDelegation?.eserviceId !== agreement.eserviceId
+  ) {
+    throw operationNotAllowed(authData.organizationId);
   }
 };
 
-export const assertRequesterCanSuspend = assertRequesterCanActivate;
```

Comment:
Diff is misleading here, I actually removed `assertRequesterCanActivate` and `assertRequesterCanSuspend`